### PR TITLE
Resize! support

### PR DESCRIPTION
**Pull request by matinraayai**
Adds support for resize! for ROCVector.

### CODE REVIEW

File: src/array.jl [hunk 1]
```diff
@@ -519,3 +519,37 @@ end
...
+      maxsize
+    end
+    new_buf = Mem.alloc(A.buf.device, bufsize)
+    copy_size = min(length(A), n) * sizeof(T)
+    if copy_size > 0
+      Mem.transfer!(new_buf, A.buf, copy_size)
+    end
+    Mem.retain(new_buf)
+    if Mem.release(A.buf)
+      Mem.free(A.buf)
```
Comment (jpsamaroo) [inline src/array.jl:549]: Missed indentation here

File: src/array.jl [hunk 2]
```diff
@@ -519,3 +519,37 @@ end
...
+    bufsize = if Base.isbitsunion(T)
+      # type tag array past the data
+      maxsize + n
+    else
+      maxsize
+    end
+    new_buf = Mem.alloc(A.buf.device, bufsize)
+    copy_size = min(length(A), n) * sizeof(T)
+    if copy_size > 0
+      Mem.transfer!(new_buf, A.buf, copy_size)
```
Comment (jpsamaroo) [inline src/array.jl:545]: Indentation

File: test/rocarray/base.jl [hunk 3]
```diff
@@ -27,6 +27,31 @@ end
...
 end
 
+@testset "resize!" begin
+    a_h = Array(range(1, 10))
+    a_d = a_h |> roc
+    # Resize up
+    resize!(a_h, 15)
+    resize!(a_d, 15)
+    a_h[10:15] .= 15
+    a_d[10:15] .= 15
```
Comment (jpsamaroo) [inline test/rocarray/base.jl:37]: I think you meant to check that `a_h[1:10] .== Array(a_d[1:10])`? The appended bytes are uninitialized, so we can't predict their values.

File: test/rocarray/base.jl [hunk 4]
```diff
@@ -27,6 +27,31 @@ end
...
+    @allowscalar begin
+        @test a_h == a_d
```
Comment (jpsamaroo) [inline test/rocarray/base.jl:39]: This won't work due to uninitialized data, you can just remove this test.

File: test/rocarray/base.jl [hunk 5]
```diff
@@ -27,6 +27,31 @@ end
...
+    end
+    # Keep the size as is
+    resize!(a_h, 15)
+    resize!(a_d, 15)
+    @allowscalar begin
+        @test a_h == a_d
```
Comment (jpsamaroo) [inline test/rocarray/base.jl:45]: As above, test `a_h[1:10] .== Array(a_d[1:10])`

File: test/rocarray/base.jl [hunk 6]
```diff
@@ -27,6 +27,31 @@ end
...
+    end
+    # Resize down
+    resize!(a_h, 3)
+    resize!(a_d, 3)
+    @allowscalar begin
+        @test a_h == a_d
```
Comment (jpsamaroo) [inline test/rocarray/base.jl:51]: As above, test `a_h[1:10] .== Array(a_d[1:10])`

File: test/rocarray/base.jl [hunk 7]
```diff
@@ -27,6 +27,31 @@ end
     @test collect(c) == 3:4
 end
 
+@testset "resize!" begin
```
Comment (jpsamaroo) [inline test/rocarray/base.jl:30]: Can you also add a check, for each of the 3 resize kinds, that `length(a_h) == length(a_d)`?

**Review by jpsamaroo — COMMENTED** [review]
Great stuff!

File: test/rocarray/base.jl (hunk 3 above)
Comment (matinraayai) [reply test/rocarray/base.jl:37]: I wanted to write them with the same value to check if we can write to it, and to initialize those new bytes to have predictable values.

File: test/rocarray/base.jl (hunk 7 above)
Comment (matinraayai) [reply test/rocarray/base.jl:30]: Even though normal equality between abstract arrays will check for that, I will write a line to check for the length too.

File: test/rocarray/base.jl (hunk 4 above)
Comment (matinraayai) [reply test/rocarray/base.jl:39]: See above comment. I initialize the new elements to have the same value on both host and device. Therefore the equality should pass.

File: test/rocarray/base.jl (hunk 6 above)
Comment (matinraayai) [reply test/rocarray/base.jl:51]: Since both arrays will have the length 3, there is no need for slicing.

File: test/rocarray/base.jl (hunk 3 above)
Comment (jpsamaroo) [reply test/rocarray/base.jl:38]: Right, I misread this, sorry!

File: test/rocarray/base.jl (hunk 5 above)
Comment (jpsamaroo) [reply test/rocarray/base.jl:47]: Nevermind!

**Review by jpsamaroo — APPROVED** [review]
LGTM! Just waiting on CI.

**Comment by jpsamaroo** [comment]
Thanks!

**Pull request merged**